### PR TITLE
MSC0000: Dynamic room predecessor

### PR DESCRIPTION
**Pull request by MadLittleMods**
A proposal for specifying a room's predecessor after the fact.

Rendered (todo)

### CODE REVIEW

File: proposals/3946-dynamic-predecessor.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+This allows the predecessor state to be updated as necessary.
+
+Event type: State event
+State key: A zero-length string.
+
+**`m.room.predecessor` event `content` field definitions:**
+
+key | type | value | description | required
+--- | --- | --- | --- | ---
+`predecessor_room_id` | string | Room ID | A reference to the room that came before and was replaced by this room | yes
```
Comment (MadLittleMods) [inline proposals/3946-dynamic-predecessor.md:44]: `predecessor_room_id` is a bit wordy but plainly describes what it is and what the value should be 🤷

File: proposals/3946-dynamic-predecessor.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+would require extensive [authorization
+rule](https://spec.matrix.org/v1.5/rooms/v10/#authorization-rules) changes and
+considerations.
+
+It would also be good to consider whether `predecessor` should be part of the
+`m.room.create` at all. Maybe we should consider removing it in favor of this new event
+type.
+
+
+## Security considerations
```
Comment (dkasak) [inline proposals/3946-dynamic-predecessor.md:67]: There's another thing to consider here. If we allow a subsequent state event to change the predecessor, then a state reset can potentially reset it to an earlier value, or even make different homeservers observe different values. Moreover, this can happen even without a state reset in case of conditions such as a network split.

Obviously we should eradicate state resets, making this much rarer, but it's somewhat intrinsic to the MSC that situations like this can happen, as soon as we allow dynamically changing the predecessor.

Perhaps this isn't a problem, but I think it's worth mentioning.

Comment (turt2live) [inline proposals/3946-dynamic-predecessor.md:67]: I'm a bit worried this MSC introduces higher risk of security issues in clients regarding room takeovers. Currently, the relationship between the previous and current room can be verified (baring state resets), giving the client confidence that room B is in fact the most recent edition of room A. With the relationship being mutable, clients are likely to not check the predecessor correctly and therefore allow someone to hide/take over non-tombstoned rooms.

The current relationship of rooms is very linear, and though this MSC says the mutable event takes precedence, not all clients will have updated to consider that point. This can lead to inconsistencies in what users see, or abusive scenarios where rooms are hidden from some people and not others.

Comment (dkasak) [reply proposals/3946-dynamic-predecessor.md:67]: (I think we race conditioned a bit and share [more-or-less the same concern](https://github.com/matrix-org/matrix-spec-proposals/pull/3946#discussion_r1062703513), just worded a bit differently.)

Comment (turt2live) [reply proposals/3946-dynamic-predecessor.md:67]: I'm slightly more concerned about clients getting the implementation wrong, despite warnings in the spec, though yes: the root issue is "mutable is scary", I think.

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:67]: Resolving in favor of the other thread, https://github.com/matrix-org/matrix-spec-proposals/pull/3946#discussion_r1062714169

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:67]: > Currently, the relationship between the previous and current room can be verified (baring state resets), giving the client confidence that room B is in fact the most recent edition of room A.

There is nothing "verifiable" about the `m.room.create` `predecessor` value besides that it came from someone with proper power levels. A Matrix room can created with whatever predecessor desired (even multiple rooms with the same predecessor). And same for the tombstone from some old room pointing to a new room. And as a note in case onlookers aren't as familiar, tombstones are already mutable.

To me, the chain of tombstone and predecessor are mere suggestions of where you can continue looking at history in the room. I'm not quite following the line of thought to get the severity with danger in any of this.

All of these problems like selectively sharing the state only to certain servers or state resets seem completely separate to this MSC. If we can't trust the state in a room, then everything falls apart.

Comment (turt2live) [reply proposals/3946-dynamic-predecessor.md:67]: The presence of a tombstone pointing to the new room and a predecessor pointing to the tombstoned room forms a verifiable relationship - this was specifically designed in for room upgrades. 

With this MSC, we'd be breaking that relationship we previously went out of our way to make verifiable - this is where the concern is coming from. If we're breaking the verifiable nature of the relationship, it needs to be for justifiable/secure reasons. 

We already acknowledge that state resets and similar can cause the relationship to be broken, however the expectation in those cases is that clients actually split the room into two logical rooms again, even if half the relationship still exists. With both halves of the relationship, the client can merge the rooms together safely.

Comment (turt2live) [reply proposals/3946-dynamic-predecessor.md:67]: Reference:
image

In this diagram, if a user is joined to all 3 rooms then what should happen is Room B and Room A are treated as a single room (Room B, specifically), and Room C is treated as its own room. With this MSC, the risk becomes that Room C "takes over" the relationship, thus making all 3 rooms represented by Room C, despite Room B never intending for that to happen.

State resets can also do this, but are a known problem and indeed irrelevant as far as this discussion is concerned.

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:67]: Are there notes somewhere in the spec or wherever about what you're describing as "verifiable". From usage, it doesn't appear that way to me at all.

I can already create Room C with the predecessor of Room B without this MSC. Nothing seems "verifiable" about that.

`POST /_matrix/client/v3/createRoom`

```json
{
    "creation_content": {
        "predecessor": "!RFPsWIrSoCdxfOivXv:my.matrix.host"
    },
    "preset": "public_chat",
    "name": "predecessor_test1"
}
```

And this MSC doesn't change the linear linked-list relationship with tombstones and predecessors that already exists.

Comment (turt2live) [reply proposals/3946-dynamic-predecessor.md:67]: It's possible to create such a room, just as it's possible to create a tombstone event that goes nowhere (or a room avatar with invalid media reference, or a message without `body`, etc). The current spec is incredibly vague because it was written in an era where that happened more than we would have liked: https://spec.matrix.org/v1.5/client-server-api/#client-behaviour-27

There might be some history in https://github.com/matrix-org/matrix-spec-proposals/pull/1791 (the spec PR) as it looks like there was some back and forth on the PR (because that was also legal in the spec process at the time), and the original MSC was https://github.com/matrix-org/matrix-spec-proposals/pull/1502 (note that it's technically MSC1501, for reasons of process at the time). There's also more discussion at https://github.com/matrix-org/matrix-spec-proposals/issues/1425 which might be useful.

I haven't looked in depth, but it appears the sentence you're after isn't described explicitly, but rather implicitly. Because the client is expected to make an effort to hide the old room from the user, it is also expected to ensure malicious predecessor structures the user has visibility on do not override rooms (as that would be a takeover). 

The js-sdk's implementation of this verification is togglable, as it somewhat depends on the scenario where it's being used: https://github.com/matrix-org/matrix-js-sdk/blob/424212cd6546a00b606380f10ae2585d1d43e518/src/client.ts#L5015

Similar filtering is done by the js-sdk when determining the true set of "visible" rooms (used by the room list algorithms in the react-sdk layer): https://github.com/matrix-org/matrix-js-sdk/blob/424212cd6546a00b606380f10ae2585d1d43e518/src/client.ts#L3768

In short, the verification happens client-side at the moment (which is regrettable, but there's much more inferred history to room upgrades that made this reality - I can dig into it internally if interested, though it doesn't affect this MSC).

----

fwiw, some of this concern is driven by the MSC's problem statement: it's not entirely clear what exact problem is being solved here, and some out of band poking has revealed a few more details about the real-world case for this. In practice, it looks like Gitter will be migrating chat history into Matrix and would like to do that in a clean way. The proposed solution appears to be that a MSC2716-compatible room will be created, history imported, tombstoned, and retroactive predecessor set on the already-existing room to point to the MSC2716-compatible room.

Assuming that's accurate (please correct me if not), there are a few concerns/questions/suggestions regarding the approach:
1. In order to guarantee any amount of durability on the imported history, Gitter would need to be using a stable room version to import the history into. Currently, MSC2716 is unstable and Synapse drops unstable room versions when the feature is included in a stable version, intentionally and regardless of usage. This will mean that eventually, when/if MSC2716 gets adopted into a stable room version, the history will be lost to time. It also means the history would only be accessible by Synapse currently, as no other server to my knowledge implements the unstable room version.
2. MSC2716 has a backwards compatibility case to allow importing history into existing rooms, as the room creator. From what I remember, all of the Gitter rooms were created by a service account: this service account could be used to import history into the already-existing room version, therefore not requiring a second room to be involved.
3. If the Gitter service account doesn't have creator access in enough of the rooms, but the Gitter server can upgrade the room, an option would be to upgrade the room in-place to a MSC2716-supported version (ideally stable, for reasons described in no. 1 above), import the history into the *new* room, and continue using the room normally (or upgrade it again for a clean slate): this gives the benefit of time-accurate history (not necessarily DAG-accurate, which clients don't care about anyways) while also requiring no changes to room upgrades. 

The value of a dynamic predecessor is definitely present, I think, though the security concern of room takeovers has me worried. Having written this short thesis of a comment, I don't really have many good ideas on how to fix it right now, but believe it can be done. In the meantime, hopefully one of the suggestions/considerations for the real-world usecase are useful.

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:67]: > In short, the verification happens client-side at the moment (which is regrettable, but there's much more inferred history to room upgrades that made this reality - I can dig into it internally if interested, though it doesn't affect this MSC).

I don't see a problem where this MSC conflicts with any of this. The client can still do its merging and "verification" as necessary. And the client getting it wrong is not a dire scenario. It just results in two rooms being shown.


> fwiw, some of this concern is driven by the MSC's problem statement: it's not entirely clear what exact problem is being solved here, and some out of band poking has revealed a few more details about the real-world case for this. 

The MSC describes the import use case that I will be using this for but it's not with MSC2716.

For Gitter, we're using the single `/send?ts=123` endpoint with timestamp massaging. We're creating a separate historical room, importing the message history one by one, tombstoning to point to the "live" room and want to use this MSC to have the "live" room's predecessor point back to the historical room.

Comment (turt2live) [reply proposals/3946-dynamic-predecessor.md:67]: > > In short, the verification happens client-side at the moment (which is regrettable, but there's much more inferred history to room upgrades that made this reality - I can dig into it internally if interested, though it doesn't affect this MSC).
> 
> I don't see a problem where this MSC conflicts with any of this. The client can still do its merging and "verification" as necessary. And the client getting it wrong is not a dire scenario. It just results in two rooms being shown. This "verification" is even toggleable as you mentioned so it's not core to anything.

A client getting it wrong means the room has been taken over, which is very much against our threat model for Matrix (for better or worse). This MSC allows for dynamic predecessors, but clients trying to apply their verification logic to the linked list will end up treating the proposed usecase as two isolated rooms instead of a single room. 



> > fwiw, some of this concern is driven by the MSC's problem statement: it's not entirely clear what exact problem is being solved here, and some out of band poking has revealed a few more details about the real-world case for this.
> 
> The MSC describes the import use case that I will be using this for but it's [not with MSC2716](https://gitlab.com/gitterHQ/webapp/-/merge_requests/2313#why-arent-we-using-msc2716).
> 
> For Gitter, we're using the single `/send?ts=123` endpoint with timestamp massaging. We're creating a separate historical room, importing the message history one by one, tombstoning to point to the "live" room and want to use this MSC to have the "live" room's predecessor point back to the historical room.

If this could be clarified in the MSC, it would be appreciated. The current intention of the MSC is unclear and appears to tell a different story than what is actually planned.

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:67]: > A client getting it wrong means the room has been taken over, which is very much against our threat model for Matrix (for better or worse).

How is this a takeover? Above all, this is just a client display thing. And second, it looks like one room needs the tombstone and other one needs a predecessor for things to look replaced. Both sides acknowledged each other for this to happen. Anything other than this seems like a bad assumption on the clients part.

> If this could be clarified in the MSC, it would be appreciated. The current intention of the MSC is unclear and appears to tell a different story than what is actually planned.

I can update it to spell it out more plainly 👍

File: proposals/3946-dynamic-predecessor.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
+# MSC3946: Dynamic room predecessor
+
+Currently, the only way to specify the predecessor of a room is during room creation in
+the `m.room.create` state event which can't be changed after the room is created
+(the `m.room.create` event is immutable and can't be changed retrospectively).
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:5]: I'm not sure this is wrong 🤔 (more opinions/knowledge needed)

File: proposals/3946-dynamic-predecessor.md (hunk 2 above)
Comment (turt2live) [reply proposals/3946-dynamic-predecessor.md:73]: > How is this a takeover? Above all, this is just a client display thing. And second, it looks like one room needs the tombstone and other one needs a predecessor for things to look replaced. Both sides acknowledged each other for this to happen. Anything other than this seems like a bad assumption on the clients part.

The case I'm concerned about isn't when both rooms point at each other, but rather when a client developer inevitably makes a mistake in the implementation of this MSC. There's already significant risk of clients getting the implementation wrong, and making the predecessor dynamic opens the door wider for mistakes (despite warnings in the spec, which often go unheeded). This would result in a malicious room being allowed to overwrite/hide another room it shouldn't have right to, thus takeover.

File: proposals/3946-dynamic-predecessor.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (turt2live) [reply proposals/3946-dynamic-predecessor.md:5]: "retroactively" is more correct as the subject is being looked at without consideration.

File: proposals/3946-dynamic-predecessor.md (hunk 2 above)
Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:73]: > This would result in a malicious room being allowed to overwrite/hide another room it shouldn't have right to, thus takeover.

This seems like it would be a client doing something very wrong. We can't help a client that doesn't look at both `tombstone` and `predecessor` to see that they match. That's a bare minimum even with the current state of `tombstone`/`predecessor` if you're making rooms look replaced. A client can show the pointer forward/backward without this verification though.

Even a naive client implementation that only looks at the create event for the `predecessor` is safe in the new world that this MSC creates.

If a client implementation updates to look at `m.room.predecessor`, then that is also pretty fool-proof regardless of which era of state they look at. The only tricky thing I could think of is if we also lived in a [MSC3779](https://github.com/matrix-org/matrix-spec-proposals/pull/3779) world where someone doesn't check for the canoncial version with `state_key: ""` but that's not a concern yet.

File: proposals/3946-dynamic-predecessor.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:5]: We seem to use `retrospectively` more often in the spec than `retroactively` in scenarios like this (just searching all of the specs in this repo)

Judging from https://english.stackexchange.com/a/148988/103469, it seems like `retroactively` fits better since an action is being taken to change things.

File: proposals/3946-dynamic-predecessor.md (hunk 2 above)
Comment (turt2live) [reply proposals/3946-dynamic-predecessor.md:73]: I agree that the client would be doing something wrong, however the risk of a client doing something wrong is higher with an MSC like this (in its current form) that overall leaves an impression that the protocol is insecure, even if the blame is actually with the clients. 

Unfortunately I don't recall all of the details, but there was also a reason why we embedded the predecessor into the create event instead of a dedicated state event: a dedicated state event would have been cleaner and "more proper" as far as Matrix is concerned, but we landed on putting it in the create event for a reason. Digging up that rationale would likely help this conversation, as it keeps coming to mind, though I sadly do not have the bandwidth to do that research.

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:73]: > I agree that the client would be doing something wrong, however the risk of a client doing something wrong is higher with an MSC like this (in its current form) that overall leaves an impression that the protocol is insecure, even if the blame is actually with the clients.

I don't think this MSC increases the risk level. We have the exact same risk with the current `tombstone`/`predecessor` state of things.

And to address that one aspect, it sounds we should consider another MSC to verify this server-side so clients can more easily rely on eating whatever we feed it.


> Unfortunately I don't recall all of the details, but there was also a reason why we embedded the predecessor into the create event instead of a dedicated state event: a dedicated state event would have been cleaner and "more proper" as far as Matrix is concerned, but we landed on putting it in the create event for a reason. Digging up that rationale would likely help this conversation, as it keeps coming to mind, though I sadly do not have the bandwidth to do that research.

This would be interesting to find. We should start a new thread if we come up with anything here.

Comment (andybalaam) [reply proposals/3946-dynamic-predecessor.md:73]: One datapoint around verifiability: I found a bug in matrix-js-sdk's code for verifying links: https://github.com/matrix-org/matrix-js-sdk/pull/3089 . The fact that this has remained undiscovered for a long time may imply that people are not paying a great deal of attention to verifiability.

File: proposals/3946-dynamic-predecessor.md (hunk 1 above)
Comment (andybalaam) [inline proposals/3946-dynamic-predecessor.md:50]: @MadLittleMods did you consider including event_id like in the create event? Some code in matrix-react-sdk expects to be able to have that. I'm not sure yet what the implications would be of not providing it when looking for a room's predecessor. (I imagine we can work around it, but I haven't looked into it yet)

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:50]: > did you consider including event_id like in the create event?

Didn't know about it so no consideration was given to it so far. For my own reference: https://spec.matrix.org/v1.5/client-server-api/#mroomcreate

Adding a `last_known_event_id` field seems fine 👍

I assume it's not related to the upgrade verification part at all since we should be looking at the latest state in the room. But this sort of info could be nice as one option for a continuation point to jump to in the old room.

It terms of workaround fallbacks, we can just go to where the latest tombstone in the old room is or just the latest in the old room. One more complicated option is tor use the `/timestamp_to_event` endpoint to find the latest event from when the predecessor was sent but doesn't seem necessary or the right thing to do exactly.

Comment (t3chguy) [reply proposals/3946-dynamic-predecessor.md:50]: Without a via list this is likely to be unjoinable over federation as room_ids are not routable. https://github.com/vector-im/element-web/issues/2925 for context.

**Comment by justjanne** [comment]
One desired use case with this is to allow users to view and access a predecessor room they haven't been in before, and which their homeserver likely doesn't know about.

@t3chguy Just raised the issue with me that to join such rooms, the room id isn't enough, we'd also need the vias or a room alias. Which means the vias should be included at least in the dynamic predecessor message. As the static room predecessor message doesn't include vias either, use of that likely should be phased out.

File: proposals/3946-dynamic-predecessor.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+## Proposal
+
+Add a `m.room.predecessor` state event type where you can specify the room predecessor.
+This allows the predecessor state to be updated as necessary.
+
+Event type: State event
+State key: A zero-length string.
+
+**`m.room.predecessor` event `content` field definitions:**
```
Comment (justjanne) [inline proposals/3946-dynamic-predecessor.md:46]: One desired use case with this is to allow users to view and access a predecessor room they haven't been in before, and which their homeserver likely doesn't know about.

@t3chguy Just raised the issue with me that to join such rooms, the room id isn't enough, we'd also need the vias or a room alias. Which means the vias should be included at least in the dynamic predecessor message. As the static room predecessor message doesn't include vias either, use of that likely should be phased out.

Comment (justjanne) [reply proposals/3946-dynamic-predecessor.md:46]: (As requested, I've moved the comment to a separate thread)

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:46]: > Without a via list this is likely to be unjoinable over federation as room_ids are not routable. https://github.com/vector-im/element-web/issues/2925 for context.
>
> @t3chguy, https://github.com/matrix-org/matrix-spec-proposals/pull/3946/files#r1100243871

(moving this comment to this thread)

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:46]: This makes a lot of sense! 👍

Added a `via_servers` field to the `m.room.predecessor` event in the proposal

File: proposals/3946-dynamic-predecessor.md (hunk 1 above)
Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md:50]: So I understand the use cases, are my assumptions above correct? Are there other use cases? Is this useful?

File: proposals/3946-dynamic-predecessor.md (hunk 4 above)
Comment (t3chguy) [reply proposals/3946-dynamic-predecessor.md:46]: IMO this should be `via` for consistency with already specced shapes, e.g. `m.space.child`. https://spec.matrix.org/v1.5/client-server-api/#mspacechild

File: proposals/3946-dynamic-predecessor.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+key | type | value | description | required
+--- | --- | --- | --- | ---
+`predecessor_room_id` | string | Room ID | A reference to the room that came before and was replaced by this room | yes
+`via_servers` | [string] | List of server names | The servers to attempt to join the room through. These servers should be participating in the room to be useful. This option is necessary since room ID's are not routable on their own.
```
Comment (t3chguy) [inline proposals/3946-dynamic-predecessor.md:51]: IMO this should be `via` for consistency with already specced shapes, e.g. `m.space.child`. https://spec.matrix.org/v1.5/client-server-api/#mspacechild

File: proposals/3946-dynamic-predecessor.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+The presence of `m.room.predecessor` state in the room should take priority over the
+`predecessor` specified in the `m.room.create` event.
+
+An additional note is that since via servers are not specified alongside the
+`predecessor` in a `m.room.create` event, the `m.room.create` usage should probably be
+phased out in favor of this new event.
```
Comment (justjanne) [inline proposals/3946-dynamic-predecessor.md:58]: The reasoning for this is that, even if the via servers were specified in the room.create event, they'd quickly go out of date, as the room.create event is immutable.

Whenever via servers change, the dynamic room predecessor event can be updated with new via information, though.

As result, even if the room.create event were updated to include via servers, phasing it out would still be recommended.

Comment (MadLittleMods) [reply proposals/3946-dynamic-predecessor.md]: Added 👍

Thanks for the extra notes!

**Comment by nyabinary** [comment]
So whats the status of this?
